rpcs3/Emu/Cell/PPUThread.cpp

```diff
@@ -485,19 +485,89 @@ extern void ppu_register_function_at(u32 addr, u32 size, u64 ptr)
 	return ppu_register_function_at(addr, size, reinterpret_cast<ppu_function_t>(ptr));
 }
 
+u32 ppu_get_exported_func_addr(u32 fnid, const std::string& module_name);
+
 struct ppu_far_jumps_t
 {
-	std::unordered_map<u32, std::pair<u32, bool>> vals;
+	struct all_info_t
+	{
+		u32 target;
+		bool link;
+		bool with_toc;
+		std::string module_name;
+	};
+
+	std::unordered_map<u32, all_info_t> vals;
 
 	mutable shared_mutex mutex;
 
-	std::pair<u32, bool> get_target(u32 pc) const
+	// Get target address, 'ppu' is used in ppu_far_jump in order to modify registers
+	u32 get_target(const u32 pc, ppu_thread* ppu = nullptr)
 	{
 		reader_lock lock(mutex);
 
 		if (auto it = vals.find(pc); it != vals.end())
 		{
-			return it->second;
+			all_info_t& all_info = it->second;
+			u32 target = all_info.target;
+
+			bool link = all_info.link;
+			bool from_opd = all_info.with_toc;
+
+			if (!all_info.module_name.empty())
+			{
+				target = ppu_get_exported_func_addr(target, all_info.module_name);
+			}
+
+			if (from_opd && !vm::check_addr<sizeof(ppu_func_opd_t)>(target))
+			{
+				// Avoid reading unmapped memory under mutex
+				from_opd = false;
+			}
+
+			if (from_opd)
+			{
+				auto& opd = vm::_ref<ppu_func_opd_t>(target);
+				target = opd.addr;
+
+				// We modify LR to custom values here 
+				link = false;
+
+				if (ppu)
+				{
+					// This check checks if we need to either branch and save modified registers
+					// *** Notice how this check is not compatible with using the same call twice without any other in between! ***
```

Comment:
Hmm. Can you elaborate on why call to next is handled in a special manner and what does this comment mean, that simple recursion is not allowed?

Reply:
Recursion is not recommended but there are cases in which it's valid here such as if the HLE calls history looks like this:

1. [caller CIA] jumpf funcA
2. [funcA + 0x10] jumpf funcA // funcA + 0x10 != caller CIA (comparing 1.CIA and 2.CIA)
3. [caller CIA] jumpf funcA // Same (comparing 2.CIA and 3.CIA)

Reply:
How about using PPU stack with artificial call frames? It seems it could look more simple in the end, and easier to track the callstack.

Reply:
It is a callstack, that's not the issue. Ok I think I need to reserve a special address for registers restoration callback, its address will be passed in LR. Using the same address for registers restoration was probably a mistake.

Reply:
Fixed.